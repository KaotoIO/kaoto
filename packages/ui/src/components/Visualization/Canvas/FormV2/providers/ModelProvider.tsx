import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

export interface ModelContextValue {
  model: unknown;
  errors?: Record<string, string[]>;
  onPropertyChange: (propName: string, value: unknown) => void;
  disabled?: boolean;
}

export const ModelContext = createContext<ModelContextValue>({ model: {}, onPropertyChange: () => {} });

export const ModelContextProvider: FunctionComponent<PropsWithChildren<ModelContextValue>> = ({
  model,
  errors,
  onPropertyChange,
  disabled,
  children,
}) => {
  const value = useMemo(() => {
    return { model, errors, onPropertyChange, disabled };
  }, [model, errors, onPropertyChange, disabled]);

  return <ModelContext.Provider value={value}>{children}</ModelContext.Provider>;
};
