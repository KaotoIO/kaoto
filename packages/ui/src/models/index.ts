export * from './camel-catalog-index';
export * from './camel-components-catalog';
export * from './camel-processors-catalog';
export * from './catalog-kind';
export * from './flow';
export * from './kamelets-catalog';
export * from './schema';
