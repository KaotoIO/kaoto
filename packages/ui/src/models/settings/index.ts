export * from './default-settings-adapter';
export * from './settings.model';
