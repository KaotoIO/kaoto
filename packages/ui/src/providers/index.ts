export * from './catalog-schema-loader.provider';
export * from './entities.provider';
