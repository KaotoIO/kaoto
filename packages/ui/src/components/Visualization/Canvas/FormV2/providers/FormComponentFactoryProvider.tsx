import { createContext, FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { StringField } from '../fields/StringField';
import { PasswordField } from '../fields/PasswordField';
import { FieldProps } from '../typings';

type FormComponentFactoryContextValue = (schema: KaotoSchemaDefinition['schema']) => FunctionComponent<FieldProps>;

export const FormComponentFactoryContext = createContext<FormComponentFactoryContextValue | undefined>(undefined);

export const FormComponentFactoryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const factory = useCallback<FormComponentFactoryContextValue>((schema) => {
    if (schema.format === 'password') {
      return PasswordField;
    }
    switch (schema.type) {
      case 'string':
      case 'number':
      case 'integer':
        return StringField;
      case 'boolean':
        return BooleanField;
      case 'object':
        return ObjectField;
      case 'array':
        return () => <div>Array field</div>;
    }

    if (Array.isArray(schema.oneOf)) {
      return OneOfField;
    } else if (Array.isArray(schema.anyOf)) {
      return () => <div>AnyOf field</div>;
    }

    return DisabledField;
  }, []);

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};
