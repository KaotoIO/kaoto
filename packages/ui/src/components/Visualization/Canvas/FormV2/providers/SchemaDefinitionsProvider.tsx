import { createContext, FunctionComponent, PropsWithChildren } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';

type SchemaDefinitionsContextValue = Record<string, KaotoSchemaDefinition['schema']>;

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>
> = ({ schema, children }) => {
  const definitions: SchemaDefinitionsContextValue = schema.definitions ?? {};

  return <SchemaDefinitionsContext.Provider value={definitions}>{children}</SchemaDefinitionsContext.Provider>;
};
