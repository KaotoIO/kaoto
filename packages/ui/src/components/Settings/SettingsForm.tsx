import { Button, Card, CardBody, CardFooter, CardTitle } from '@patternfly/react-core';
import { FunctionComponent, useCallback, useContext, useState } from 'react';
import settingsSchema from '../../assets/settingsSchema.json';
import { useReloadContext } from '../../hooks/useReloadContext/useReloadContext';
import { KaotoSchemaDefinition } from '../../models';
import { SettingsModel } from '../../models/settings';
import { SchemaBridgeProvider } from '../../providers/schema-bridge.provider';
import { SettingsContext } from '../../providers/settings.provider';
import { CustomAutoForm } from '../Form/CustomAutoForm';

export const SettingsForm: FunctionComponent = () => {
  const settingsAdapter = useContext(SettingsContext);
  const { lastRender, reloadPage } = useReloadContext();
  const [settings, setSettings] = useState(settingsAdapter.getSettings());

  const onChangeModel = useCallback((value: unknown) => {
    setSettings(value as SettingsModel);
  }, []);

  const onSave = useCallback(() => {
    settingsAdapter.saveSettings(settings);
    reloadPage();
  }, [reloadPage, settings, settingsAdapter]);

  return (
    <Card data-last-render={lastRender}>
      <CardTitle>Settings</CardTitle>

      <CardBody>
        <SchemaBridgeProvider schema={settingsSchema as KaotoSchemaDefinition['schema']}>
          <CustomAutoForm data-testid="settings-form" model={settings} onChangeModel={onChangeModel} />
        </SchemaBridgeProvider>
      </CardBody>

      <CardFooter>
        <Button data-testid="settings-form-save-btn" variant="primary" onClick={onSave}>
          Save
        </Button>
      </CardFooter>
    </Card>
  );
};
