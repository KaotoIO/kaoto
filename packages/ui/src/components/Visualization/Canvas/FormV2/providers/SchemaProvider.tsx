import { createContext, FunctionComponent, PropsWithChildren, useContext, useMemo } from 'react';
import { KaotoSchemaDefinition } from '../../../../../models';
import { resolveSchemaWithRef } from '../../../../../utils';
import { SchemaDefinitionsContext } from './SchemaDefinitionsProvider';

export interface SchemaContextValue {
  schema: KaotoSchemaDefinition['schema'];
  definitions: Record<string, KaotoSchemaDefinition['schema']>;
}

export const SchemaContext = createContext<SchemaContextValue>({ schema: {}, definitions: {} });

export const SchemaProvider: FunctionComponent<PropsWithChildren<{ schema: KaotoSchemaDefinition['schema'] }>> = ({
  schema,
  children,
}) => {
  const { definitions, omitFields } = useContext(SchemaDefinitionsContext);

  const value = useMemo(() => {
    const resolvedSchema = resolveSchemaWithRef(schema, definitions);

    if (Array.isArray(resolvedSchema.anyOf)) {
      resolvedSchema.anyOf = resolvedSchema.anyOf.map((anyOfSchema) => resolveSchemaWithRef(anyOfSchema, definitions));
    }
    if (Array.isArray(resolvedSchema.oneOf)) {
      resolvedSchema.oneOf = resolvedSchema.oneOf.map((oneOfSchema) => resolveSchemaWithRef(oneOfSchema, definitions));
    }

    omitFields.forEach((field) => {
      delete resolvedSchema.properties?.[field];
    });

    return { schema: resolvedSchema, definitions };
  }, [definitions, omitFields, schema]);

  return <SchemaContext.Provider value={value}>{children}</SchemaContext.Provider>;
};
