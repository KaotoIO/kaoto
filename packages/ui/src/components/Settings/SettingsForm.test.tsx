import { act, fireEvent, render, screen } from '@testing-library/react';
import { AbstractSettingsAdapter, DefaultSettingsAdapter } from '../../models/settings';
import { ReloadContext, SettingsProvider } from '../../providers';
import { SettingsForm } from './SettingsForm';
import { MemoryRouter } from 'react-router-dom';

describe('SettingsForm', () => {
  let reloadPage: jest.Mock;
  let settingsAdapter: AbstractSettingsAdapter;

  const wrapper = ({ children }: { children: React.ReactNode }) => {
    return (
      <MemoryRouter>
        <ReloadContext.Provider value={{ reloadPage, lastRender: 0 }}>
          <SettingsProvider adapter={settingsAdapter}>{children}</SettingsProvider>
        </ReloadContext.Provider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    reloadPage = jest.fn();
    settingsAdapter = new DefaultSettingsAdapter();
    render(<SettingsForm />, { wrapper });
  });

  it('should render', () => {
    expect(screen.getByTestId('settings-form')).toMatchSnapshot();
  });

  it('should update settings upon clicking save', () => {
    act(() => {
      const input = screen.getByLabelText('Camel Catalog URL');
      fireEvent.change(input, { target: { value: 'http://localhost:8080' } });
    });

    act(() => {
      const button = screen.getByTestId('settings-form-save-btn');
      fireEvent.click(button);
    });

    expect(settingsAdapter.getSettings().catalogUrl).toBe('http://localhost:8080');
  });

  it('should not update settings if the save button was not clicked', () => {
    act(() => {
      const input = screen.getByLabelText('Camel Catalog URL');
      fireEvent.change(input, { target: { value: 'http://localhost:8080' } });
    });

    expect(settingsAdapter.getSettings().catalogUrl).not.toBe('http://localhost:8080');
  });

  it('should reload the page upon clicking save', () => {
    act(() => {
      const button = screen.getByTestId('settings-form-save-btn');
      fireEvent.click(button);
    });

    expect(reloadPage).toHaveBeenCalledTimes(1);
  });
});
