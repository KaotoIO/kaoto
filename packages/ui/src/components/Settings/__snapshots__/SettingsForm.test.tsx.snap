// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm should render 1`] = `
<form
  autocomplete="new-password"
  class="pf-v5-c-form"
  data-testid="settings-form"
  novalidate=""
>
  <div>
    <div
      class="pf-v5-c-form__group"
      data-fieldname="catalogUrl"
      data-testid="wrapper-field"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="uniforms-0000-0002"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            Camel Catalog URL
          </span>
        </label>
         
        <div
          style="display: contents;"
        >
          <button
            aria-disabled="false"
            aria-label="More info for field"
            class="pf-v5-c-button pf-m-plain field-hint-button"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            data-testid="field-hint-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <span
          class="pf-v5-c-form-control"
        >
          <input
            aria-invalid="false"
            aria-label="uniforms text field"
            data-ouia-component-id="OUIA-Generated-TextInputBase-1"
            data-ouia-component-type="PF5/TextInput"
            data-ouia-safe="true"
            data-testid="text-field"
            description="URL that points to the \`index.json\` file from the Kaoto Camel Catalog. Leave it empty to use the embedded catalog."
            id="uniforms-0000-0002"
            label="Camel Catalog URL"
            name="catalogUrl"
            type="text"
            value=""
          />
        </span>
        <div
          class="pf-v5-c-form__helper-text"
        >
          <div
            class="pf-v5-c-helper-text"
          >
            <div
              class="pf-v5-c-helper-text__item"
            >
              <span
                class="pf-v5-c-helper-text__item-text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      data-fieldname="nodeLabel"
      data-testid="wrapper-field"
      options="[object Object],[object Object]"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="uniforms-0000-0005"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            Node label to display in canvas
          </span>
        </label>
         
        <div
          style="display: contents;"
        >
          <button
            aria-disabled="false"
            aria-label="More info for field"
            class="pf-v5-c-button pf-m-plain field-hint-button"
            data-ouia-component-id="OUIA-Generated-Button-plain-2"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            data-testid="field-hint-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <div
          class="pf-v5-c-menu-toggle pf-m-full-width pf-m-typeahead"
        >
          <div
            class="pf-v5-c-text-input-group pf-m-plain"
          >
            <div
              autocomplete="off"
              class="pf-v5-c-text-input-group__main"
              data-testid="create-typeahead-select-input"
              id="create-typeahead-select-input"
            >
              <span
                class="pf-v5-c-text-input-group__text"
              >
                <input
                  aria-controls="select-create-typeahead-listbox"
                  aria-expanded="false"
                  aria-label="Type to filter"
                  class="pf-v5-c-text-input-group__text-input"
                  placeholder="Select an option"
                  role="combobox"
                  type="text"
                  value="description"
                />
              </span>
            </div>
            <div
              class="pf-v5-c-text-input-group__utilities"
            >
              <button
                aria-disabled="false"
                aria-label="Clear input value"
                class="pf-v5-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-3"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 352 512"
                  width="1em"
                >
                  <path
                    d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <button
            aria-expanded="false"
            aria-label="Menu toggle"
            class="pf-v5-c-menu-toggle__button"
            data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-1"
            data-ouia-component-type="PF5/MenuToggle"
            data-ouia-safe="true"
            tabindex="-1"
            type="button"
          >
            <span
              class="pf-v5-c-menu-toggle__controls"
            >
              <span
                class="pf-v5-c-menu-toggle__toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-form__group"
      data-fieldname="nodeToolbarTrigger"
      data-testid="wrapper-field"
      options="[object Object],[object Object]"
    >
      <div
        class="pf-v5-c-form__group-label"
      >
        <label
          class="pf-v5-c-form__label"
          for="uniforms-0000-0008"
        >
          <span
            class="pf-v5-c-form__label-text"
          >
            Open Node toolbar
          </span>
        </label>
         
        <div
          style="display: contents;"
        >
          <button
            aria-disabled="false"
            aria-label="More info for field"
            class="pf-v5-c-button pf-m-plain field-hint-button"
            data-ouia-component-id="OUIA-Generated-Button-plain-4"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            data-testid="field-hint-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="pf-v5-c-form__group-control"
      >
        <div
          class="pf-v5-c-menu-toggle pf-m-full-width pf-m-typeahead"
        >
          <div
            class="pf-v5-c-text-input-group pf-m-plain"
          >
            <div
              autocomplete="off"
              class="pf-v5-c-text-input-group__main"
              data-testid="create-typeahead-select-input"
              id="create-typeahead-select-input"
            >
              <span
                class="pf-v5-c-text-input-group__text"
              >
                <input
                  aria-controls="select-create-typeahead-listbox"
                  aria-expanded="false"
                  aria-label="Type to filter"
                  class="pf-v5-c-text-input-group__text-input"
                  placeholder="Select an option"
                  role="combobox"
                  type="text"
                  value="onHover"
                />
              </span>
            </div>
            <div
              class="pf-v5-c-text-input-group__utilities"
            >
              <button
                aria-disabled="false"
                aria-label="Clear input value"
                class="pf-v5-c-button pf-m-plain"
                data-ouia-component-id="OUIA-Generated-Button-plain-5"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 352 512"
                  width="1em"
                >
                  <path
                    d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  />
                </svg>
              </button>
            </div>
          </div>
          <button
            aria-expanded="false"
            aria-label="Menu toggle"
            class="pf-v5-c-menu-toggle__button"
            data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-2"
            data-ouia-component-type="PF5/MenuToggle"
            data-ouia-safe="true"
            tabindex="-1"
            type="button"
          >
            <span
              class="pf-v5-c-menu-toggle__controls"
            >
              <span
                class="pf-v5-c-menu-toggle__toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-card pf-m-flat custom-nest-field"
      data-ouia-component-id="OUIA-Generated-Card-2"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      data-testid="nest-field"
      description="Enable / disable experimental features"
      id="uniforms-0000-000b"
      properties="[object Object]"
    >
      <div
        class="pf-v5-c-card__header"
      >
        <div
          class="pf-v5-c-card__header-main"
        >
          <div
            class="pf-v5-c-card__title"
          >
            <div
              class="pf-v5-c-card__title-text"
              id="uniforms-0000-000b-title"
            >
              Experimental features
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-c-card__body"
      >
        <div
          class="pf-v5-c-form__group"
          data-testid="wrapper-field"
        >
          <div
            class="pf-v5-c-form__group-control"
          >
            <div
              style="display: flex; align-items: center;"
            >
               
              <div
                class="pf-v5-c-check"
              >
                <input
                  aria-invalid="false"
                  class="pf-v5-c-check__input"
                  data-ouia-component-id="OUIA-Generated-Checkbox-1"
                  data-ouia-component-type="PF5/Checkbox"
                  data-ouia-safe="true"
                  data-testid="bool-field"
                  id="uniforms-0000-000d"
                  name="experimentalFeatures.enableDragAndDrop"
                  type="checkbox"
                />
                <label
                  class="pf-v5-c-check__label"
                  for="uniforms-0000-000d"
                >
                  Enable Drag & Drop
                </label>
              </div>
              <div
                style="display: contents;"
              >
                <button
                  aria-disabled="false"
                  aria-label="More info for field"
                  class="pf-v5-c-button pf-m-plain field-hint-button"
                  data-ouia-component-id="OUIA-Generated-Button-plain-6"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  data-testid="field-hint-button"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-form__helper-text"
              >
                <div
                  class="pf-v5-c-helper-text"
                >
                  <div
                    class="pf-v5-c-helper-text__item"
                  >
                    <span
                      class="pf-v5-c-helper-text__item-text"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
